### PR TITLE
Adds Omnitools for engineer and medical cyborgs, reducing on inventory clutter.

### PR DESCRIPTION
**Pull request by Singul0**
## About The Pull Request
[This PR is a bounty requested by Ophaq and worked on by Singul0.](https://tgstation13.org/phpBB/viewtopic.php?f=5&t=36013)

All of the following description in this PR is written by Ophaq as to what this PR entails:
In this PR, the medical and engineering cyborg's tools are completely reworked and condensed into an arm similar to the implant a carbon would get. The tools are shown in a radial wheel around the character to quick select what is needed instead of looking for it in a cluttered bag of items. There are a few tools such as the blood filter for the medical cyborg, as well as the welder, gas analyzer, and t-ray scanner for the engineering cyborg excluded from the radial wheel. mostly due to their inherent inmodularity

image
image

Each cyborg gets two arms in case the player wishes to have one on the side to quick swap to, like having a scalpel in one arm and a hemostat in the other on the hotbar for convenience or just preference. An upgraded version of the tools has been added to each respective cyborg upgrade node with somewhat faster action speed. The upgrade replaces the arms and transforms them into the "advanced" version which is currently the same sprite as the regular but just a faster and more efficient version. The sprites for the surgical arm currently look good but may need replacing later if someone who wishes to resprite them down the line decides to do so.
## Why It's Good For The Game
As it currently stands, the medical cyborg's magical bag of gadgets takes up a lot of your screen space and as a player who plays medical A LOT, this was a MUCH NEEDED quality of life feature.

The amount of clutter in a medical cyborg's bag makes it in my opinion, hard to see at the bottom of the screen and a nuisance to constantly close compared to other models. My standard set up for playing medical cyborg on the hotbar is 1=med analyzer, 2=usually a secondary surgery tool or injector, and 3=another surgery tool. The flow of gameplay during surgery ends up being surgery tool, hit 3 and drop it, surgery tool, repeat or for efficiency using X to swap between the two surgery tools I need on 2 and 3. This gets tedious especially after so many hours of playing medical cyborg. I know some people may disagree, but I think it would help a lot of help to speed up this flow of gameplay during surgery and declutter.

By turning the medical cyborg's toolset into an omni-surgery tool which functions like the surgery arm implant's radial wheel, this would greatly declutter by like an entire row and make things easier on medical cyborg players. Having a secondary in the bag helps with efficiency for those players who like having an extra tool on their hotbar and swapping back and forth would also improve efficiency and make less swapping by hitting Z needed. Additionally with the upgraded version as an optional upgrade in the mediborg tech, this also lets them be on par with players who use advanced tools late game but not at the level of alien tools where players would obviously out compete a mediborg in terms of action speed.

Engineering models also benefit from this rework but at a slightly different and lesser way whereas certain tools are excluded such as the welder, due to the way they work on refill and the gas scanner and t-ray scanner not counting as tool components are not included in the arms.

Syndicate versions of the engineering and medical cyborg also get these arms, unupgraded.
## Changelog
:cl:
add: Adds an omnitoolset for both engineering and medical cyborgs, containing various basic tools
qol: Engineer and Medical module inventory space is now significantly decluttered
/:cl:

### CODE REVIEW

**Comment by moocowswag** [comment]
finally the weakness of borgs wont be to just attack them from the south because they cant see you through their ui.

**Comment by ophaq** [comment]
> finally the weakness of borgs wont be to just rush them from the south because they cant see you through their ui.

Their one weakness finally conquered... or is it? You still have 3 rows to squint through but it is finally less ass.

**Comment by SomeoneYouProbablyKnow** [comment]
Would it be possible to condense all the materials into a similar arm for the engi borg? Tiles, metal, rods, maybe even the RCD? Or is the stacking too much of a pain in the ass? Same thing for general chemistry tools for the medborg, beakers dropper syringe etc.

**Comment by ophaq** [comment]
> Would it be possible to condense all the materials into a similar arm for the engi borg? Tiles, metal, rods, maybe even the RCD? Or is the stacking too much of a pain in the ass? Same thing for general chemistry tools for the medborg, beakers dropper syringe etc.

This would need to be a separate PR and I am assuming if this was a thing, the use of a custom TGUI menu would be required to be able to select what materials are selected for the engineering model. As for the medborg, that too would need its own. Having a transformable dropper/syringe would be cool to save on inventory space. Beakers should stay as is and maybe if not ambitiously and possibly too strong, have a synthesizer beaker similar to what service borgs get for their drink glasses. I have a few other changes in mind that I want to do with medical cyborgs such as adding 1/5/10 unit modularity to the output of their injectors and adding in formaldehyde in their starter reagent lists due to many reasons but again, it may be a future PR that I'll request or someone else could take on.

The dream goal is getting cyborg bags down to 2 rows. I'm glad it's currently at 3.

**Review by Wallemations — APPROVED** [review]

**Comment by MrMelbert** [comment]
As I mentioned in the forum thread, you should give them two, maybe even three omnitools. 

That way if you're doing a project which necessitates going back and forth between two tools, you don't need to keep re-opening the wheel. 

Like how a human would have a scalpel in one hand and a hemostat in the other when doing surgery.

**Comment by DATA-xPUNGED** [comment]
I don't think having 2 or 3 arms is really all that needed, have any of you used a tool implant before? It is pretty seamless to swap between tools because it is on the same spot of where you're usually using them

File: code/game/objects/items/robot/items/tools.dm
Comment (san7890) [inline code/game/objects/items/robot/items/tools.dm:256]:
```suggestion
	desc = "[desc]\n" + "It seems that this one has been upgraded to perform tasks faster."
```

File: code/game/objects/items/robot/robot_upgrades.dm
Comment (san7890) [inline code/game/objects/items/robot/robot_upgrades.dm:456]: early return instead `if(!.); return .`)

File: code/game/objects/items/robot/robot_upgrades.dm
Comment (san7890) [inline code/game/objects/items/robot/robot_upgrades.dm:462]: same here, the indentation gets ugly fast

File: code/game/objects/items/robot/robot_upgrades.dm
Comment (san7890) [inline code/game/objects/items/robot/robot_upgrades.dm:475]: ditto early return

File: code/game/objects/items/shields.dm
Comment (san7890) [inline code/game/objects/items/shields.dm:74]:
```suggestion
	if(!istype(attackby_item, /obj/item/melee/baton) || !is_bashable)
		return
		
	if(!COOLDOWN_FINISHED(src, baton_bash))
		return
	user.visible_message(span_warning("[user] bashes [src] with [attackby_item]!"))
	playsound(user.loc, 'sound/effects/shieldbash.ogg', 50, TRUE)
	COOLDOWN_START(src, baton_bash, BATON_BASH_COOLDOWN)
```

early return to lessen bad-looking indentation, also that return on the end is totally superfluous

File: code/modules/research/designs/mechfabricator_designs.dm
Comment (san7890) [inline code/modules/research/designs/mechfabricator_designs.dm:1438]:
```suggestion
		RND_CATEGORY_MECHFAB_CYBORG_MODULES + RND_SUBCATEGORY_MECHFAB_CYBORG_MODULES_ENGINEERING,
```

File: code/game/objects/items/robot/items/tools.dm
Comment (san7890) [inline code/game/objects/items/robot/items/tools.dm:334]: this comment seems pretty useless as it might change over time with no one updating the comment

**Review by san7890 — CHANGES_REQUESTED** [review]
nits but also some formatting considerations

File: code/game/objects/items/shields.dm
Comment (Singul0) [reply code/game/objects/items/shields.dm:74]: wait what, baton bash? how the hell did that commit made it's way here. I really need to fix my fork

File: code/game/objects/items/robot/robot_upgrades.dm
Comment (Singul0) [reply code/game/objects/items/robot/robot_upgrades.dm:456]: I don't follow?

**Comment by ophaq** [comment]
> I like the idea partially, omni's inherent issues are the fact you constantly have to open up a pinwheel which is giong to make overall actions from this PR slower
It was way slower having to drop the tool then select a tool in the messy bag of objects the cyborg had. This to me, feels much more efficient and faster gameplay wise. 

> on Engineering downstream at least we made baby versions of advanced tools for engineering off the bat, then they get upgraded variants that match their upgraded versions organics get. Meaning you just tap Z to swap modules
>
 That is exactly how this works. The only difference is due to some issues the welder, gas scanner and t-ray had to be omitted from the arm. The upgraded version can be printed at robotics which should match the upgraded tool action speed which humans would get for their experimental tools research.
> What'd i suggest for advanced medical upgrades would be to drop the omni part and give them the advanced tools directly in the same fashion as mentioned above
This was considered but giving them from the get go would raise balance concerns and the whole point of the omni tool thing was to get the radial menu in for more precise tool selection instead of looking in a cluttered bag, in addition to just decluttering it.


> As I mentioned in the forum thread, you should give them two, maybe even three omnitools.
> 
> That way if you're doing a project which necessitates going back and forth between two tools, you don't need to keep re-opening the wheel.
> 
> Like how a human would have a scalpel in one hand and a hemostat in the other when doing surgery.
Yep that is what the cyborgs get. Two felt sufficient in the testing because even with one, it felt more smoother in tool transition speed that I even felt the second was a bit redundant, but it's there for preference and say, hemostat quick swap.


> I don't think having 2 or 3 arms is really all that needed, have any of you used a tool implant before? It is pretty seamless to swap between tools because it is on the same spot of where you're usually using them
Like above, the arm felt smooth to transition to other tools with. The second is there for preference for players who want it. If people feel that it is too redundant later on after testing by the general player base, the second -could- be taken out but that is another PR and argument for another day in the future.

File: code/game/objects/items/robot/items/tools.dm
Comment (MrMelbert) [reply code/game/objects/items/robot/items/tools.dm:256]:
```suggestion
	desc += "\nIt seems that this one has been upgraded to perform tasks faster."
```
Radical third party suggestion

**Comment by Singul0** [comment]
^ oh and also to cancel surgeries, you need two omnitools

**Comment by mc-oofert** [comment]
so instead of just clicking the relevant tool in the menu you now have to find the omnitool and also find your relevant tool in a radial menu

i dont think thats convenient

**Comment by Kubisopplay** [comment]
Engineering borgs should have upgraded tools by default to keep parity with current status. Locking yet another thing behind a hard to get research is not cool

**Comment by DATA-xPUNGED** [comment]
Some of yall have never used a tools implant before and it shows

**Comment by mc-oofert** [comment]
> Some of yall have never used a tools implant before and it shows

i have and they sucked a lot

**Comment by GuillaumePrata** [comment]
An alternative to this system that could work (and I failed to code in the past when I was a baby coder) would be a single toolbox item in the borgs' inventory.

You try to add the toolbox to your hand and it asks what tool you want with the radial menu, then the tool goes to your hand while the toolbox stays there in the inventory.

**Comment by GuillaumePrata** [comment]
Complicated code wise, yes.

But it is better UX than having 2 tools arms that you need to constantly cycle around from a tool to another and back depending on what you are doing.

Ignoring that the toolbox is an universal item that everyone will understand how it works from prior knowledge, the cyborg arm is an unique icon that doesn't convey much and I'm sure a lot of people will be confused the first time they join as Borg because of this if they don't read PRs.

**Comment by Kubisopplay** [comment]
Current engiborg tools are twice as fast as normal human ones, and faster in fact than powertools. Could you please get them to parity so it is not a stealth nerf?

**Comment by Singul0** [comment]
> Current engiborg tools are twice as fast as normal human ones, and faster in fact than powertools. Could you please get them to parity so it is not a stealth nerf?

Fixed, the base speed should be the same as before now.

**Comment by Dawnseer** [comment]
I'd absolutely love these omnitools as much as I love toolset implants. As for the people who think they're hard to use, stop using your mouse so much and start using your empty hand inhand. It's so easy and smooth.

**Comment by Zergspower** [comment]
> > I like the idea partially, omni's inherent issues are the fact you constantly have to open up a pinwheel which is giong to make overall actions from this PR slower
> 
> It was way slower having to drop the tool then select a tool in the messy bag of objects the cyborg had. This to me, feels much more efficient and faster gameplay wise.
> That is exactly how this works. The only difference is due to some issues the welder, gas scanner and t-ray had to be omitted from the arm. The upgraded version can be printed at robotics which should match the upgraded tool action speed which humans would get for their experimental tools research.

Tapping Z or your action button to swap to the secondary tool (think advanced medical tools) is far faster is why i mentioned it

> This was considered but giving them from the get go would raise balance concerns and the whole point of the omni tool thing was to get the radial menu in for more precise tool selection instead of looking in a cluttered bag, in addition to just decluttering it.
What i meant by baby tools is just a normal toolset but in the same mechanical style as advanced, then the upgrade gives them the advanced tools.

**Comment by carlarctg** [comment]
> I'd absolutely love these omnitools as much as I love toolset implants. As for the people who think they're hard to use, stop using your mouse so much and start using your empty hand inhand. It's so easy and smooth.

I added this because I'm super awesome.
Maybe it should be added to cyborgs as well, actually. Having the third hand slot function like that would fix most issues.

**Comment by ophaq** [comment]
> > I'd absolutely love these omnitools as much as I love toolset implants. As for the people who think they're hard to use, stop using your mouse so much and start using your empty hand inhand. It's so easy and smooth.
> 
> I added this because I'm super awesome. Maybe it should be added to cyborgs as well, actually. Having the third hand slot function like that would fix most issues.

This PR literally functions like this. You use your slot and the radial menu appears after pressing Z. No need to add anything, It functions exactly like the implants. This PR should be good to go unless some new current PR is gonna conflict with it, which there shouldn't be.

**Comment by ophaq** [comment]
Since #82493 got closed, I asked to revert cyborgs back to what their previous state was for tool speed back to speed_mod. That way they still have what they had before as that surgery speed PR wont (yet) affect anything until it's reopened if it does.

**Comment by ophaq** [comment]
Nevermind it was originally a bug fix. PR is good to go then.

**Comment by optimumtact** [comment]
this is worse ux

**Comment by ophaq** [comment]
> this is worse ux

What do you mean? What's ux?

**Comment by optimumtact** [comment]
user experience.

**Comment by ophaq** [comment]
> user experience.

Did you try it out? It's way better than current standard experience where you click the tool, drop it, move to next tool. Often times you end up clicking the wrong tool too in the cluttered as all hell bag of stuff. In my opinion and testing, this improved the way medical cyborgs are able to select medical tools by a lot and it is way more convenient than the current way they have to select their tools for surgery.

**Comment by Dorsisdwarf** [comment]
How do radial tools work for people who don't use the hotkey mode?

**Comment by DATA-xPUNGED** [comment]
> How do radial tools work for people who don't use the hotkey mode?

This means that you can't press Z to interact, right? I'd imagine that the best way to use it then is to move the action button closer to the middle of the screen

**Comment by Dawnseer** [comment]
Yeah probably that. I do that for the modsuit module menu.

**Comment by Singul0** [comment]
guh, review please

File: code/game/objects/items/robot/items/tools.dm
Comment (jlsnow301) [inline code/game/objects/items/robot/items/tools.dm:229]: are you expecting more?

File: code/game/objects/items/robot/items/tools.dm
Comment (jlsnow301) [inline code/game/objects/items/robot/items/tools.dm:257]: autodoc comment pls

File: code/game/objects/items/robot/items/tools.dm
Comment (jlsnow301) [inline code/game/objects/items/robot/items/tools.dm:268]: same here

File: code/game/objects/items/robot/items/tools.dm
Comment (jlsnow301) [inline code/game/objects/items/robot/items/tools.dm:189]: Tool speed shouldnt get a comment (should be commented on parent)
reference should

File: code/game/objects/items/robot/items/tools.dm
Comment (Singul0) [reply code/game/objects/items/robot/items/tools.dm:229]: ah, it's just as an example item on how to add new items. should probably add another tool beside that

Comment (jlsnow301) [reply code/game/objects/items/robot/items/tools.dm:229]: should add `SHOULD_CALL_PARENT(FALSE)` if youre just using it as a reference

**Comment by ophaq** [comment]
Oh is this why this didn't get merged? I thought it was because the maintainers were very busy and forgot. I was just about to ask them to merge this too lol.

**Comment by Singul0** [comment]
the merge conflict wasn't there couple days ago, weird how the bot didn't detect it though

File: code/game/objects/items/robot/items/tools.dm
Comment (Singul0) [reply code/game/objects/items/robot/items/tools.dm:229]: > should add `SHOULD_CALL_PARENT(FALSE)` if youre just using it as a reference

gotcha, where should I add `SHOULD_CALL_PARENT(FALSE)`? i'm not too familiar with spacemanDMM

Comment (jlsnow301) [reply code/game/objects/items/robot/items/tools.dm:230]: sorry I shouldve been more specific, it goes here in the parent, ideally top of the proc. this prevents other people from returning ..() etc, which looks to not be what you want here

Comment (Singul0) [reply code/game/objects/items/robot/items/tools.dm:230]: ah, that's what that does. I get it now. thanks for describing it! it's added now

File: code/game/objects/items/robot/items/tools.dm
Comment (jlsnow301) [inline code/game/objects/items/robot/items/tools.dm:321]: this is not the parent proc
```suggestion
```

File: code/game/objects/items/robot/items/tools.dm
Comment (jlsnow301) [inline code/game/objects/items/robot/items/tools.dm:373]:
```suggestion
```

**Comment by ophaq** [comment]
OK so there are a few issues that were oversights that can be fixed in another PR I guess but currently the bonesetter doesn't work for medical borgs in surgeries such as compound fractures and every tool is thinking it is drapes, letting you start surgeries with them. You can't attack with tools as a mediborg due to this even on harm mode.

**Pull request merged**
